### PR TITLE
[Sparse] GCN Example

### PR DESCRIPTION
**Pull request by paoxiaode**
## Description

The GCN example code https://github.com/dmlc/dgl/issues/4931 using sparse matrix API

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Performace
|dataset | loss | val acc| test acc|
| - | - |  -|  -|
| Cora | 0.237 | 0.788 | 0.810 |

### CODE REVIEW

File: examples/sparse/gcn.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+from dgl.data import CoraGraphDataset
+from dgl.mock_sparse import create_from_coo, diag, identity
+from torch.optim import Adam
+
+class GCN(nn.Module):
+    def __init__(self, A, in_size, out_size, hidden_size=16):
+        super().__init__()
+        # two-layer GCN
```
Comment (frozenbugs) [inline examples/sparse/gcn.py:15]: Capitalize T and end with dot.

File: examples/sparse/gcn.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        self.Theta1 = nn.Linear(in_size, hidden_size)
+        self.Theta2 = nn.Linear(hidden_size, out_size)
+        self.dropout = nn.Dropout(0.5)
+        # calculate the symmetrically normalized adjacency matrix
```
Comment (frozenbugs) [inline examples/sparse/gcn.py:19]: ditto.

File: examples/sparse/gcn.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        I = identity(A.shape, device=dev)
```
Comment (frozenbugs) [inline examples/sparse/gcn.py:20]: dev is not defined?
I would suggest to calculate the norm in the main function and pass it in directly?

File: examples/sparse/gcn.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        X = F.relu(X)
+        X = self.A_hat @ self.Theta2(self.dropout(X))
+        return X
+        
+        
+def evaluate(g, pred):
+    label = g.ndata["label"]
+    val_mask = g.ndata["val_mask"]
+    test_mask = g.ndata["test_mask"]
+    # Compute accuracy on validation/test set.
```
Comment (frozenbugs) [inline examples/sparse/gcn.py:37]: Add a blank line to give it some space to breathe?

File: examples/sparse/gcn.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        super().__init__()
+        # two-layer GCN
+        self.Theta1 = nn.Linear(in_size, hidden_size)
+        self.Theta2 = nn.Linear(hidden_size, out_size)
+        self.dropout = nn.Dropout(0.5)
+        # calculate the symmetrically normalized adjacency matrix
+        I = identity(A.shape, device=dev)
+        A_hat = A + I
+        D_hat = diag(A_hat.sum(1)) ** -0.5
+        A_hat = D_hat @ A_hat @ D_hat
```
Comment (frozenbugs) [inline examples/sparse/gcn.py:23]: A_hat in the paper refer to A=I, maybe use norm here.

**Comment by frozenbugs** [comment]
What's the test result? paste it in the PR description.

File: examples/sparse/gcn.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from dgl.data import CoraGraphDataset
+from dgl.mock_sparse import create_from_coo, diag, identity
+from torch.optim import Adam
+
+class GCN(nn.Module):
+    def __init__(self, A, in_size, out_size, hidden_size=16):
```
Comment (mufeili) [inline examples/sparse/gcn.py:13]: I prefer passing `A` in `forward` to be consistent with other examples.

File: examples/sparse/gcn.py (hunk 3 above)
Comment (paoxiaode) [reply examples/sparse/gcn.py]: change to calculate the norm in the main function

File: examples/sparse/gcn.py (hunk 5 above)
Comment (paoxiaode) [reply examples/sparse/gcn.py]: rename it to A_norm

File: examples/sparse/gcn.py (hunk 6 above)
Comment (frozenbugs) [reply examples/sparse/gcn.py]: Can you also address this comment?

**Comment by frozenbugs** [comment]
@dgl-bot

File: examples/sparse/gcn.py (hunk 6 above)
Comment (paoxiaode) [reply examples/sparse/gcn.py]: I change the train function into train(g, model, A, X), but don't actually use it

**Comment by paoxiaode** [comment]
@frozenbugs can you help me to check why the checks have failed？

File: examples/sparse/gcn.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+class GCN(nn.Module):
+    def __init__(self, in_size, out_size, hidden_size=16):
+        super().__init__()
+        
+        # Two-layer GCN.
+        self.Theta1 = nn.Linear(in_size, hidden_size)
+        self.Theta2 = nn.Linear(hidden_size, out_size)
+        
+    def forward(self, A_norm, X):
```
Comment (frozenbugs) [inline examples/sparse/gcn.py:20]: Add highlight to where sparse API was used.
example:
```
    ###########################################################################
    # (HIGHLIGHT) Take the advantage of DGL sparse APIs to ...
    ###########################################################################
```

File: examples/sparse/gcn.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    g = dataset[0].to(dev)
+    num_classes = dataset.num_classes
+    X = g.ndata["feat"]
+    
+    # Create the adjacency matrix of graph.
+    src, dst = g.edges()
+    N = g.num_nodes()
+    A = create_from_coo(dst, src, shape=(N, N))
+    
+    # Calculate the symmetrically normalized adjacency matrix.
```
Comment (frozenbugs) [inline examples/sparse/gcn.py:86]: ditto.

**Review by frozenbugs — APPROVED** [review]

File: examples/sparse/gcn.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    num_classes = dataset.num_classes
+    X = g.ndata["feat"]
+
+    # Create the adjacency matrix of graph.
+    src, dst = g.edges()
+    N = g.num_nodes()
+    A = create_from_coo(dst, src, shape=(N, N))
+
+    ###########################################################
+    # (HIGHLIGHT) Compute the symmetrically normalized adjacency matrix with Sparse Matrix API
```
Comment (frozenbugs) [inline examples/sparse/gcn.py:92]: Less than 80 chars.

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by paoxiaode** [comment]
@dgl-bot
